oauth2client/contrib/multiprocess_file_storage.py

```diff
@@ -0,0 +1,355 @@
+            :class:`oauth2client.client.Credentials`.
+    """
+    data = {'file_version': 2, 'credentials': {}}
+
+    for key, credential in iteritems(credentials):
+        credential_json = credential.to_json()
+        encoded_credential = _helpers._from_bytes(base64.b64encode(
+            _helpers._to_bytes(credential_json)))
+        data['credentials'][key] = encoded_credential
+
+    credentials_file.seek(0)
+    json.dump(data, credentials_file)
+    credentials_file.truncate()
+
+
+class _MultiprocessStorageBackend(object):
+    """Thread-local backend for multiprocess storage.
+
+    Each process has only one instance of this backend per file. All threads
+    share a single instance of this backend. This ensures that all threads
+    use the same thread lock and process lock when accessing the file.
+    """
+
+    def __init__(self, filename):
+        self._file = None
+        self._filename = filename
+        self._process_lock = fasteners.InterProcessLock(
+            '{}.lock'.format(filename))
+        self._thread_lock = threading.Lock()
+        self._read_only = False
+        self._credentials = {}
+
+    def _load_credentials(self):
+        """(Re-)loads the credentials from the file."""
+        if not self._file:
+            return
+
+        loaded_credentials = _load_credentials_file(self._file)
+        self._credentials.update(loaded_credentials)
+
+        logger.debug('Read credential file')
+
+    def _write_credentials(self):
+        if self._read_only:
+            logger.debug('In read-only mode, not writing credentials.')
+            return
+
+        _write_credentials_file(self._file, self._credentials)
+        logger.debug('Wrote credential file {}.'.format(self._filename))
+
+    def acquire_lock(self):
+        self._thread_lock.acquire()
+        locked = self._process_lock.acquire(timeout=INTERPROCESS_LOCK_DEADLINE)
+
+        if locked:
+            _create_file_if_needed(self._filename)
+            self._file = open(self._filename, 'r+')
+            self._read_only = False
+
+        else:
+            logger.warn(
+                'Failed to obtain interprocess lock for credentials. '
+                'If a credential is being refreshed, other processes may '
+                'not see the updated access token and refresh as well.')
+            if os.path.exists(self._filename):
+                self._file = open(self._filename, 'r')
+            else:
+                self._file = None
+            self._read_only = True
+
+        self._load_credentials()
+
+    def release_lock(self):
+        if self._file is not None:
+            self._file.close()
+            self._file = None
+
+        if not self._read_only:
+            self._process_lock.release()
+
+        self._thread_lock.release()
+
+    def _refresh_predicate(self, credentials):
+        if credentials is None:
+            return True
+        if credentials.invalid:
```

Comment:
done.

On Sun, Jul 10, 2016 at 9:34 AM Nathaniel Manista <notifications@github.com>
wrote:

> In oauth2client/contrib/multiprocess_file_storage.py
> <https://github.com/google/oauth2client/pull/504#discussion_r70184708>:
>
> > +        self._load_credentials()
> > +
> > +    def release_lock(self):
> > +        if self._file is not None:
> > +            self._file.close()
> > +            self._file = None
> > +
> > +        if not self._read_only:
> > +            self._process_lock.release()
> > +
> > +        self._thread_lock.release()
> > +
> > +    def _refresh_predicate(self, credentials):
> > +        if credentials is None:
> > +            return True
> > +        if credentials.invalid:
>
> At least turn these additional ifs into elifs and put the return False at
> the bottom inside an else.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/oauth2client/pull/504/files/17117ddca9367bc0773839d6e0278988a3ec6e20#r70184708>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAPUc9AY-TKJtBJO-yT979g9CLLZziQoks5qUR8BgaJpZM4IN9JU>
> .
>